src/index.test.ts

```diff
@@ -1,6 +1,1431 @@
+          primaryType,
+          message,
+          types,
+          'V3',
+        ).toString('hex'),
+      ).toThrow(`Cannot read property 'name' of null`);
+    });
+
+    it('should encode data with a custom type property set to undefined', function () {
+      const types = {
+        Person: [
+          { name: 'name', type: 'string' },
+          { name: 'wallet', type: 'address' },
+        ],
+        Mail: [
+          { name: 'from', type: 'Person' },
+          { name: 'to', type: 'Person' },
+          { name: 'contents', type: 'string' },
+        ],
+      };
+      const primaryType = 'Mail';
+      const message = {
+        from: {
+          name: 'Cow',
+          wallet: '0xCD2a3d9F938E13CD947Ec05AbC7FE734Df8DD826',
+        },
+        contents: 'Hello, Bob!',
+      };
+
+      expect(
+        sigUtil.TypedDataUtils.encodeData(
+          primaryType,
+          message,
+          types,
+          'V3',
+        ).toString('hex'),
+      ).toMatchSnapshot();
+    });
+
+    it('should throw an error when trying to encode a function', function () {
+      const types = {
+        Message: [{ name: 'data', type: 'function' }],
+      };
+      const message = { data: 'test' };
+
+      expect(() =>
+        sigUtil.TypedDataUtils.encodeData(
+          'Message',
+          message,
+          types,
+          'V3',
+        ).toString('hex'),
+      ).toThrow('Unsupported or invalid type: function');
+    });
+
+    it('should throw an error when trying to encode with a missing primary type definition', function () {
+      const types = {};
+      const message = { data: 'test' };
+
+      expect(() =>
+        sigUtil.TypedDataUtils.encodeData(
+          'Message',
+          message,
+          types,
+          'V3',
+        ).toString('hex'),
+      ).toThrow('No type definition specified: Message');
+    });
+
+    it('should throw an error when trying to encode an unrecognized type', function () {
+      const types = {
+        Message: [{ name: 'data', type: 'foo' }],
+      };
+      const message = { data: 'test' };
+
+      expect(() =>
+        sigUtil.TypedDataUtils.encodeData(
+          'Message',
+          message,
+          types,
+          'V3',
+        ).toString('hex'),
+      ).toThrow('Unsupported or invalid type: foo');
+    });
+
+    it('should encode data when given extraneous types', function () {
+      const types = {
+        Message: [{ name: 'data', type: 'string' }],
+        Extra: [{ name: 'data', type: 'string' }],
+      };
+      const message = { data: 'Hello!' };
+
+      expect(
+        sigUtil.TypedDataUtils.encodeData(
+          'Message',
+          message,
+          types,
+          'V3',
+        ).toString('hex'),
+      ).toMatchSnapshot();
+    });
+  });
+
+  // This test suite should include all expected V4 behavior except the example
+  // data encoding, whcih is covered in the "V3/V4 encoding similarities" test
+  // suite
```

Comment:
This comment was leftover from an earlier draft of this PR where some V3/V4 encoding was covered by the similarities test suite. That isn't the case anymore - only the similarities are covered there now. All behaviour for V3 and V4 are covered in their respective test suites. This comment can be deleted.

```suggestion
```